File: sdks/node/test/core/dispatch-order.test.ts
// S25 — opt-in LIFO dispatch order. Jobs are enqueued before the worker starts,
// so all pile up pending; a concurrency-1 worker then dispatches them one at a
// time and we record the order. UUIDv7 ids make same-millisecond ties
// deterministic, so LIFO is exactly newest-first.

import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, expect, it } from "vitest";
import { Queue, type Worker } from "../../src/index";

let worker: Worker | undefined;

afterEach(() => {
  worker?.stop();
  worker = undefined;
});

function newQueue(): Queue {
  const dbPath = join(mkdtempSync(join(tmpdir(), "taskito-order-")), "queue.db");
  return new Queue({ dbPath });
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

async function runAndRecord(queue: Queue, total: number): Promise<number[]> {
  const order: number[] = [];
  queue.task("rec", (n: number) => {
    order.push(n);
  });
  for (let i = 0; i < total; i++) queue.enqueue("rec", [i]);

  worker = queue.runWorker({ concurrency: 1, batchSize: 1 });
  const deadline = Date.now() + 20_000;
  while (Date.now() < deadline && order.length < total) {
    await sleep(50);
  }
  return order;
}

it("LIFO dispatches newest-first", async () => {
  const queue = newQueue();
  queue.configureQueue("default", { dispatchOrder: "lifo" });
  const order = await runAndRecord(queue, 6);
  expect(order).toEqual([5, 4, 3, 2, 1, 0]);
});

it("FIFO is the default", async () => {
  const queue = newQueue();
  const order = await runAndRecord(queue, 6);
  expect(order).toEqual([0, 1, 2, 3, 4, 5]);
});

File: sdks/java/src/test/java/org/byteveda/taskito/core/DispatchOrderTest.java
package org.byteveda.taskito.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.nio.file.Path;
import java.time.Duration;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import org.byteveda.taskito.Taskito;
import org.byteveda.taskito.task.Task;
import org.byteveda.taskito.worker.Worker;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.io.TempDir;

/**
 * S25 — opt-in LIFO dispatch order. Jobs are enqueued before the worker starts,
 * so they pile up pending; a concurrency-1 worker then dispatches them one at a
 * time and we record the order. UUIDv7 ids make same-millisecond ties
 * deterministic, so LIFO is exactly newest-first.
 */
class DispatchOrderTest {

    @Test
    void validatesOrder(@TempDir Path dir) {
        try (Taskito queue =
                Taskito.builder().url(dir.resolve("v.db").toString()).open()) {
            assertThrows(IllegalArgumentException.class, () -> queue.dispatchOrder("default", "sideways"));
        }
    }

    private static List<Integer> runAndRecord(Taskito queue, int total) throws Exception {
        List<Integer> order = new CopyOnWriteArrayList<>();
        Task<Integer> rec = Task.of("rec", Integer.class);
        for (int i = 0; i < total; i++) {
            queue.enqueue(rec, i);
        }
        try (Worker worker = queue.worker()
                .concurrency(1)
                .batchSize(1)
                .handle(rec, order::add)
                .start()) {
            long deadline = System.nanoTime() + Duration.ofSeconds(20).toNanos();
            while (System.nanoTime() < deadline && order.size() < total) {
                Thread.sleep(50);
            }
        }
        return order;
    }

    @Test
    @Timeout(40)
    void lifoDispatchesNewestFirst(@TempDir Path dir) throws Exception {
        try (Taskito queue =
                Taskito.builder().url(dir.resolve("lifo.db").toString()).open()) {
            queue.dispatchOrder("default", "lifo");
            List<Integer> order = runAndRecord(queue, 6);
            assertEquals(List.of(5, 4, 3, 2, 1, 0), order, "LIFO runs newest-first");
        }
    }

    @Test
    @Timeout(40)
    void fifoIsTheDefault(@TempDir Path dir) throws Exception {
        try (Taskito queue =
                Taskito.builder().url(dir.resolve("fifo.db").toString()).open()) {
            List<Integer> order = runAndRecord(queue, 6);
            assertEquals(List.of(0, 1, 2, 3, 4, 5), order, "FIFO runs oldest-first by default");
        }
    }
}
